### PR TITLE
Switch to CreateAsyncScope

### LINKED ISSUE #32314: Adopt new CreateAsyncDisposable method
See https://github.com/dotnet/runtime/pull/51840. We should update:

[ ] - MVC
[ ] - SignalR
[ ] - Blazor

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/dotnet/aspnetcore/issues/32314

### CODE REVIEW

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 1]
```diff
@@ -107,7 +107,7 @@ public override async Task OnConnectedAsync(HubConnectionContext connection)
 
         public override async Task OnDisconnectedAsync(HubConnectionContext connection, Exception? exception)
         {
-            var scope = _serviceScopeFactory.CreateScope();
+            var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (davidfowl) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:110]:
```suggestion
            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 2]
```diff
@@ -284,7 +284,7 @@ private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessag
             var methodExecutor = descriptor.MethodExecutor;
 
             var disposeScope = true;
-            var scope = _serviceScopeFactory.CreateScope();
+            var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (davidfowl) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:287]:
```suggestion
            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```

**Comment by BrennanConroy** [comment]
Might be able to remove https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/AsyncDisposableExtensions.cs

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 3]
```diff
@@ -73,7 +73,7 @@ internal partial class DefaultHubDispatcher<THub> : HubDispatcher<THub> where TH
 
         public override async Task OnConnectedAsync(HubConnectionContext connection)
         {
-            var scope = _serviceScopeFactory.CreateScope();
+            var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (davidfowl) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:76]:
```suggestion
            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 4]
```diff
@@ -660,7 +660,7 @@ private static void CloneHttpContext(HttpContext context, HttpConnectionContext
 
             CloneUser(newHttpContext, context);
 
-            connection.ServiceScope = context.RequestServices.CreateScope();
+            connection.ServiceScope = context.RequestServices.CreateAsyncScope();
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:663]: Did you change this type to `AsyncServiceScope`

File: src/Components/test/E2ETest/Tests/WebAssemblyAuthenticationTests.cs [hunk 5]
```diff
@@ -456,7 +456,7 @@ private void WaitUntilLoaded(bool skipHeader = false)
 
         private void EnsureDatabaseCreated(IServiceProvider services)
         {
-            using var scope = services.CreateScope();
+            using var scope = services.CreateAsyncScope();
```
Comment (davidfowl) [inline src/Components/test/E2ETest/Tests/WebAssemblyAuthenticationTests.cs:459]: You can revert this one

File: src/Components/WebView/WebView/src/WebViewManager.cs [hunk 6]
```diff
@@ -178,7 +178,7 @@ internal async Task AttachToPageAsync(string baseUrl, string startUrl)
             // change this if any important reason emerges.
             _currentPageContext?.Dispose();
 
-            var serviceScope = _provider.CreateScope();
+            var serviceScope = _provider.CreateAsyncScope();
```
Comment (davidfowl) [inline src/Components/WebView/WebView/src/WebViewManager.cs:181]: Did you change this `PageContext` to take `AsyncServiceScope`?

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs [hunk 7]
```diff
@@ -234,7 +234,7 @@ public WebAssemblyHost Build()
             // to configure services inside *that scope* inside their startup code, we create *both* the
             // service provider and the scope here.
             var services = _createServiceProvider();
-            var scope = services.GetRequiredService<IServiceScopeFactory>().CreateScope();
+            var scope = services.GetRequiredService<IServiceScopeFactory>().CreateAsyncScope();
 
             return new WebAssemblyHost(services, scope, Configuration, RootComponents, _persistedState);
```
Comment (davidfowl) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Is `WebAssemblyHost` `IAsyncDisposable`?

File: src/Components/WebAssembly/WebAssembly.Authentication/test/WebAssemblyAuthenticationServiceCollectionExtensionsTests.cs [hunk 8]
```diff
@@ -425,7 +425,7 @@ public void OidcProviderOptionsAndDependencies_NotResolvedFromRootScope()
 
             var host = builder.Build();
 
-            using var scope = host.Services.CreateScope();
+            using var scope = host.Services.CreateAsyncScope();
```
Comment (davidfowl) [inline src/Components/WebAssembly/WebAssembly.Authentication/test/WebAssemblyAuthenticationServiceCollectionExtensionsTests.cs:428]: Revert this?

File: src/Components/Server/src/Circuits/CircuitFactory.cs [hunk 9]
```diff
@@ -45,7 +45,7 @@ internal class CircuitFactory
             ClaimsPrincipal user,
             IComponentApplicationStateStore store)
         {
-            var scope = _scopeFactory.CreateScope();
+            var scope = _scopeFactory.CreateAsyncScope();
```
Comment (davidfowl) [inline src/Components/Server/src/Circuits/CircuitFactory.cs:48]: Where is this disposed?

File: src/Components/Components/src/OwningComponentBase.cs [hunk 10]
```diff
@@ -43,7 +43,7 @@ protected IServiceProvider ScopedServices
                     throw new ObjectDisposedException(GetType().Name);
                 }
 
-                _scope ??= ScopeFactory.CreateScope();
+                _scope ??= ScopeFactory.CreateAsyncScope();
```
Comment (davidfowl) [inline src/Components/Components/src/OwningComponentBase.cs:46]: This needs to be `AsyncServiceScope`

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs (hunk 7 above)
Comment (HaoK) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Yes it looks like it: https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs#L22

**Comment by HaoK** [comment]
>https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/AsyncDisposableExtensions.cs

@BrennanConroy looks like TestClient still relies on this in some of the signalr tests, I can move the DisposeAsync there if you want to nuke the file or ?

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 11]
```diff
@@ -73,7 +73,7 @@ internal partial class DefaultHubDispatcher<THub> : HubDispatcher<THub> where TH
 
         public override async Task OnConnectedAsync(HubConnectionContext connection)
         {
-            var scope = _serviceScopeFactory.CreateScope();
+            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:76]: either remove `await using` or remove the `try finally` below

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 12]
```diff
@@ -107,7 +107,7 @@ public override async Task OnConnectedAsync(HubConnectionContext connection)
 
         public override async Task OnDisconnectedAsync(HubConnectionContext connection, Exception? exception)
         {
-            var scope = _serviceScopeFactory.CreateScope();
+            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:110]: same here

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 13]
```diff
@@ -284,7 +284,7 @@ private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessag
             var methodExecutor = descriptor.MethodExecutor;
 
             var disposeScope = true;
-            var scope = _serviceScopeFactory.CreateScope();
+            await using var scope = _serviceScopeFactory.CreateAsyncScope();
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:287]: remove `await using`, this scope has a complex lifetime and is disposal is handled elsewhere

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 11 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:76]: remove the try finally.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs (hunk 7 above)
Comment (davidfowl) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Then you need to change this https://github.com/dotnet/aspnetcore/blob/2105ef1ceb87dcd1d7bf6afbab9b3e6868ccd3bf/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs#L24 to an `AsyncServiceScope`.

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 11 above)
Comment (HaoK) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:76]: I'll let you and @davidfowl hash this out, he suggested adding await using for these in a previous suggestion

File: src/Components/Server/src/Circuits/CircuitFactory.cs (hunk 9 above)
Comment (HaoK) [reply src/Components/Server/src/Circuits/CircuitFactory.cs:48]: CircuitHost is responsible for it, it had special case logic for AsyncDisposable, it's updated now as well

File: src/Components/Components/src/OwningComponentBase.cs [hunk 14]
```diff
@@ -43,8 +43,8 @@ protected IServiceProvider ScopedServices
                     throw new ObjectDisposedException(GetType().Name);
                 }
 
-                _scope ??= ScopeFactory.CreateScope();
-                return _scope.ServiceProvider;
+                _scope ??= ScopeFactory.CreateAsyncScope();
+                return _scope.Value.ServiceProvider;
```
[the remark this answers is not in the document]
Comment (HaoK) [reply src/Components/Components/src/OwningComponentBase.cs:47]: Its a nullable struct so compiler said no bueno without the Value

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Components/Components/src/OwningComponentBase.cs:47]: Ohh, now it's `Nullable<T>`

[the remark this answers is not in the document]
Comment (HaoK) [reply src/Components/Components/src/OwningComponentBase.cs:47]: Yeah I had to google what was going on

**Comment by HaoK** [comment]
MacOS crash:
```
Dump successfully written
The active test run was aborted. Reason: Test host process crashed
Results File: /private/tmp/helix/working/ADF10956/w/ABEE0936/e/TestResults/TestResults.xml

Attachments:
  /private/tmp/helix/working/ADF10956/w/ABEE0936/e/TestResults/e51930bb-6abf-4ab4-b285-9a03951ef59e/Sequence_657c382aa4074d9fbb246e374359cd6a.xml
Test Run Aborted with error System.Exception: One or more errors occurred.
Total tests: Unknown
     Passed: 195
 Total time: 48.1828 Seconds
 ---> System.Exception: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.Read7BitEncodedInt() in System.Private.CoreLib.dll:token 0x6005646+0x0
   at System.IO.BinaryReader.ReadString() in System.Private.CoreLib.dll:token 0x600563c+0x92
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.LengthPrefixCommunicationChannel.NotifyDataAvailable() in Microsoft.TestPlatform.CommunicationUtilities.dll:token 0x6000026+0x0
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.TcpClientExtensions.MessageLoopAsync(TcpClient client, ICommunicationChannel channel, Action`1 errorHandler, CancellationToken cancellationToken) in Microsoft.TestPlatform.CommunicationUtilit
ies.dll:token 0x6000064+0xa0
   --- End of inner exception stack trace ---.
The active Test Run was aborted because the host process exited unexpectedly. Please inspect the call stack above, if available, to get more information about where the exception originated from.
The test running when the crash occurred: 
Microsoft.AspNetCore.Mvc.Routing.MvcRouteHandlerTests.RouteAsync_FailOnNoAction_LogsCorrectValues
Microsoft.AspNetCore.Mvc.VirtualFileResultTest.ExecuteResultAsync_ThrowsFileNotFound_IfFileProviderCanNotFindTheFile
Microsoft.AspNetCore.Mvc.ModelBinding.Metadata.HasValidatorsValidationMetadataProviderTest.CreateValidationMetadata_SetsHasValidatorsToFalse_IfNoProviderReturnsTrue
Microsoft.AspNetCore.Mvc.Controllers.ServiceBasedControllerActivatorTest.Create_ThrowsIfControllerIsNotRegisteredInServiceProvider
Microsoft.AspNetCore.Mvc.Routing.ViewEnginePathTest.ResolvePath_DoesNotTraversePastApplicationRoot
Microsoft.AspNetCore.Mvc.Filters.ActionFilterAttributeTests.ActionFilterAttribute_ActionFilter_SettingResult_ShortCircuits
Microsoft.AspNetCore.Mvc.ModelBinding.BindAttributeTest.BindAttribute_Include
Microsoft.AspNetCore.Mvc.ModelBinding.Binders.ComplexObjectModelBinderProviderTest.Create_ForSupportedTypes_ReturnsBinder
Microsoft.AspNetCore.Mvc.HttpOkObjectResultTest.HttpOkObjectResult_InitializesStatusCodeAndValue
Microsoft.AspNetCore.Mvc.ApiExplorer.ApiConventionMatcherTest.IsMatch_ReturnsTrue_IfParamsArrayMatchesRemainingArguments
```

I didn't update anything in mvc in this PR I don't think so doesn't look related, retrying helix run

**Comment by davidfowl** [comment]
@HaoK this ready?

**Comment by HaoK** [comment]
Sure yeah, marking for ready for review

**Review by BrennanConroy — APPROVED** [review]
SignalR side looks good

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs (hunk 7 above)
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Did you miss pushing an update with this change?

Comment (HaoK) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Oops yeah I missed that, thanks, will update this file now

Comment (HaoK) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Hrm no the file looks updated actually, https://github.com/dotnet/aspnetcore/blob/haok/asyncscope/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs#L24  assuming I'm looking at the right place

Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:239]: Ooops, might have missed it.

**Review by pranavkm — APPROVED** [review]

**Pull request merged**